sktime/clustering/partitioning/_averaging_metrics.py

```diff
@@ -0,0 +1,220 @@
+
+        cost_mat = np.zeros((max_length, max_length))
+        delta_mat = np.zeros((max_length, max_length))
+        path_mat = np.zeros((max_length, max_length), dtype=np.int8)
+
+        medoid_ind = self._approximate_medoid_index(cost_mat, delta_mat)
+        center = self.series[medoid_ind]
+
+        for _ in range(self.n_iterations):
+            center = self._DBA_update(center, cost_mat, path_mat, delta_mat)
+
+        return center
+
+    def _approximate_medoid_index(self, cost_mat, delta_mat):
+        if len(self.series) <= 50:
+            indices = range(0, len(self.series))
+        else:
+            indices = np.random.choice(range(0, len(self.series)), 50, replace=False)
+
+        medoid_ind = -1
+        best_ss = 1e20
+        for index_candidate in indices:
+            candidate = self.series[index_candidate]
+            ss = self._sum_of_squares(candidate, cost_mat, delta_mat)
+            if medoid_ind == -1 or ss < best_ss:
+                best_ss = ss
+                medoid_ind = index_candidate
+        return medoid_ind
+
+    def _sum_of_squares(self, s, cost_mat, delta_mat):
+        return sum(
+            map(lambda t: self._squared_DTW(s, t, cost_mat, delta_mat), self.series)
+        )
+
+    def _DTW(self, s, t, cost_mat, delta_mat):
+        return np.sqrt(self._squared_DTW(s, t, cost_mat, delta_mat))
+
+    def _squared_DTW(self, s, t, cost_mat, delta_mat):
+        s_len = len(s)
+        t_len = len(t)
+        self._fill_delta_mat_dtw(s, t, delta_mat)
+        cost_mat[0, 0] = delta_mat[0, 0]
+        for i in range(1, s_len):
+            cost_mat[i, 0] = cost_mat[i - 1, 0] + delta_mat[i, 0]
+
+        for j in range(1, t_len):
+            cost_mat[0, j] = cost_mat[0, j - 1] + delta_mat[0, j]
+
+        for i in range(1, s_len):
+            for j in range(1, t_len):
+                diag, left, top = (
+                    cost_mat[i - 1, j - 1],
+                    cost_mat[i, j - 1],
+                    cost_mat[i - 1, j],
+                )
+                if diag <= left:
+                    if diag <= top:
+                        res = diag
+                    else:
+                        res = top
+                else:
+                    if left <= top:
+                        res = left
+                    else:
+                        res = top
+                cost_mat[i, j] = res + delta_mat[i, j]
+        return cost_mat[s_len - 1, t_len - 1]
+
+    def _fill_delta_mat_dtw(self, center, s, delta_mat):
+        slim = delta_mat[: len(center), : len(s)]
+        np.subtract.outer(center, s, out=slim)
+        np.square(slim, out=slim)
+
+    def _DBA_update(self, center, cost_mat, path_mat, delta_mat):
+        options_argmin = [(-1, -1), (0, -1), (-1, 0)]
+        updated_center = np.zeros(center.shape)
+        n_elements = np.array(np.zeros(center.shape), dtype=int)
+        center_length = len(center)
+        for s in self.series:
```

Comment:
I would consider if you can optimise this using numpy vectorisation. 

Multiple nested loops will be slow, and i think you can definitely optimise this easily.

Reply:
good point, can I suggest that once the PR is in, this could be raised as an issue?